### PR TITLE
feat: use signed url instead of auth propagation

### PR DESCRIPTION
**Pull request by anda-ren**
## Description
- use signed url instead of auth propagation in initContainers
- use signed url instead of auth propagation in client dataset
## Modules
- [ ] UI
- [x] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [x] Others

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [x] add necessary doc

### CODE REVIEW

File: server/storage-access-layer/src/main/java/ai/starwhale/mlops/storage/fs/StorageAccessServiceFile.java [hunk 1]
```diff
@@ -134,4 +146,9 @@ public void delete(String path) throws IOException {
             }
         }
     }
+
+    @Override
+    public String signedUrl(String path, Long expTimeMillis) throws IOException {
+        return serviceProvider + "/" + path + "/" + (System.currentTimeMillis() + expTimeMillis);
```
Comment (jialeicui) [inline server/storage-access-layer/src/main/java/ai/starwhale/mlops/storage/fs/StorageAccessServiceFile.java:152]: Do we need to encrypt this info?

File: server/controller/src/main/java/ai/starwhale/mlops/api/DatasetApi.java [hunk 2]
```diff
@@ -246,6 +246,24 @@ void pullLinkContent(
             @RequestParam(name = "size", required = false) String size,
             HttpServletResponse httpResponse);
 
+    @Operation(summary = "Sign SWDS uri to get a temporarily accessible link",
+            description = "Sign SWDS uri to get a temporarily accessible link")
```
[the remark this answers is not in the document]
Comment (anda-ren) [reply server/controller/src/main/java/ai/starwhale/mlops/api/DatasetApi.java:250]: image
`temporarily` is right 
https://www.grammarcheck.net/editor/

File: client/starwhale/api/_impl/dataset/loader.py [hunk 3]
```diff
@@ -55,17 +54,23 @@ def _get_store(self, row: TabularDatasetRow) -> ObjectStore:
...
 
-        if row.object_store_type == ObjectStoreType.REMOTE:
-            _store = ObjectStore.from_data_link_uri(row.data_uri, row.auth_name)
+        _type = self.dataset_uri.instance_type
+        if _type == InstanceType.CLOUD:
+            _store = ObjectStore.from_remote_dataset(self.dataset_uri, row.auth_name)
         else:
-            _store = ObjectStore.from_dataset_uri(self.dataset_uri)
+            if row.object_store_type == ObjectStoreType.REMOTE:
+                _store = ObjectStore.from_data_link_uri(row.data_uri, row.auth_name)
```
[the remark this answers is not in the document]
Comment (anda-ren) [reply client/starwhale/api/_impl/dataset/loader.py:62]: #### `Controller has no enough bandwidth.`
If `s3` is the backend of controller, then a `s3 pre-signed link` is returned to client which directlly goes to s3 service. 

The only case the bandwidth directlly goes to controller is that the backend of controller doesn't support a `pre-signed link` like local fs.
#### `I made a dataset with the old controller and the new controller can not access the old OSS`
This means auth info lost, client can't get the auth info from controller neither. So s3 client can't work either.

File: client/starwhale/core/dataset/store.py [hunk 4]
```diff
@@ -408,6 +415,30 @@ def _make_file(self, bucket: str, key_compose: str) -> FileLikeObj:
...
+            headers = {"Range": f"bytes={_start}-{_end}"}
+        r = requests.get(url, headers=headers)
+        return io.BytesIO(r.content)
+
+    def sign_uri(self, uri: str, auth_name: str) -> t.Any:
+        r = self.do_http_request(
+            f"/project/{self.dataset_uri.project}/{self.dataset_uri.object.typ}/{self.dataset_uri.object.name}/version/{self.dataset_uri.object.version}/sign-link",
+            method=HTTPMethod.GET,
+            instance_uri=self.dataset_uri,
+            params={"uri": uri, "authName": auth_name, "expTimeMillis": 1000 * 60 * 30},
```
[the remark this answers is not in the document]
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:437]: #### 30min may not enough for a big task.
this is row level , not task level. 30 min is enough I think . User sign this url when he needs the file. Though configurable is still better.

#### And do we need expTime in milliseconds
yes, seconds or minutes is enough . However milliseconds is enough also.

File: client/starwhale/api/_impl/dataset/loader.py [hunk 5]
```diff
@@ -55,17 +54,23 @@ def _get_store(self, row: TabularDatasetRow) -> ObjectStore:
         if _store:
             return _store
 
-        if row.object_store_type == ObjectStoreType.REMOTE:
-            _store = ObjectStore.from_data_link_uri(row.data_uri, row.auth_name)
+        _type = self.dataset_uri.instance_type
+        if _type == InstanceType.CLOUD:
+            _store = ObjectStore.from_remote_dataset(self.dataset_uri, row.auth_name)
```
[the remark this answers is not in the document]
Comment (anda-ren) [reply client/starwhale/api/_impl/dataset/loader.py:59]: when `self.dataset_uri.instance_type` is `InstanceType.CLOUD` the `backend` for `DataLoader` can only be `SignedUrlBackend` no matter what the `ObjectSotreType` is. So it is not orthogonal in this case.

when `self.dataset_uri.instance_type` is `InstanceType.STANDALONE` the `backend` for `DataLoader` can be `LocalFSStorageBackend` or `S3StorageBackend`. So it is some kind of orthogonal

File: server/storage-access-layer/src/main/java/ai/starwhale/mlops/storage/fs/StorageAccessServiceFile.java (hunk 1 above)
Comment (anda-ren) [reply server/storage-access-layer/src/main/java/ai/starwhale/mlops/storage/fs/StorageAccessServiceFile.java:152]: no need for now i think

File: client/starwhale/core/dataset/store.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jialeicui) [reply client/starwhale/core/dataset/store.py:437]: `_get_store` has a cache

https://github.com/star-whale/starwhale/blob/17bee36d95ecd97afcdd7bef38c72782782509b7/client/starwhale/api/_impl/dataset/loader.py#L52-L56

File: server/storage-access-layer/src/main/java/ai/starwhale/mlops/storage/fs/StorageAccessServiceFile.java (hunk 1 above)
Comment (jialeicui) [reply server/storage-access-layer/src/main/java/ai/starwhale/mlops/storage/fs/StorageAccessServiceFile.java:152]: User can get others data when controller uses local fs.

Comment (anda-ren) [reply server/storage-access-layer/src/main/java/ai/starwhale/mlops/storage/fs/StorageAccessServiceFile.java:152]: if user use local fs in production he will suffer a lot of other issues besides security

File: client/starwhale/core/dataset/store.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:437]: how about fix the cache key from `object_store_type.auth_name` to `dataset_uri.auth_name`

File: server/controller/src/main/java/ai/starwhale/mlops/schedule/k8s/K8sTaskScheduler.java [hunk 6]
```diff
@@ -209,19 +206,41 @@ private Map<String, String> buildCoreContainerEnvs(Task task) {
...
+                    log.error("sign model url failed for {} ", path, e);
+                    throw new SwProcessException(ErrorType.STORAGE).tip("sign model url failed");
+                }
+            });
+        } catch (IOException e) {
+            log.error("list model files failed for {} ", swJob.getSwmp().getPath(), e);
+            throw new SwProcessException(ErrorType.STORAGE).tip("list model files failed");
+        }
+        try {
+            storageAccessService.list(jobRuntime.getStoragePath()).forEach(path -> {
```
[the remark this answers is not in the document]
Comment (anda-ren) [reply server/controller/src/main/java/ai/starwhale/mlops/schedule/k8s/K8sTaskScheduler.java:226]: as this implemention would change  according to #1440 #1035 . let's resolve this conversation now.

**Review by jialeicui — APPROVED** [review]
LGTM

**Comment by anda-ren** [comment]
merge to #1443

**Pull request closed without merging**